### DISCUSSION TITLE
Inject vs Put

### DISCUSSION THREAD

**Discussion by quekyj** [intel/rohd#328] [Q&A]
When is a suitable time to use `inject()` and `put()`? Any rules of thumb that can be followed? It's seems to be that `put()` alone is enough for testing.

**Comment by mkorbel1** [comment]
First, to clarify the difference between the two:
- `put` places a value on a signal immediately
- `inject` is equivalent to calling `put` inside of a `Simulator.injectAction` call

A `put` on a signal will propagate changes to any combinational logic "instantly" and synchronously, without any involvement from the `Simulator`.  An `inject` will only register the `put` to occur at the next opportunity at the current `Simulator.time` while the `Simulator` is actively running.  The implication is that anything relying on events from the `Simulator` to behave properly will behave differently on `put` vs `inject`.

One key place this is different is on any sequential logic (`Sequential`, `FlipFlop`, etc.).  The `Simulator` must be involved to properly understand how to sample transitions that happen before vs. after a clock edge.  Modifying a clock into sequential logic with `put` will not trigger an edge, but modifying a clock during a running simulation will trigger a clock edge.  If you're modifying a clock or input to sequential logic during a simulation, you probably want to use `inject`.

Another case is in testbench or model logic interacting with hardware.  There's often code like `await clk.nextPosedge` in these cases, and you often want to drive some change at the current timestep during simulation upon those events.  In these cases, `inject` makes sense.

A third example is the `WaveDumper`, which relies on signals `changed` event, which is calculated using the `Simulator` phases.  If you just call `put`, depending on when it's called, the `WaveDumper` may not record the signal change in the proper timestamp in the waves.

I think a good rule of thumb is:
- If you're modelling combinational logic (e.g. custom gate) or interacting with combinational logic, then `put` is convenient and you don't need to use the `Simulator` at all.
- If you're running the `Simulator` and want to interact with a design over simulation time, you probably want to use `inject`.  Especially if you're collecting waves, have sequential logic, or are modelling stateful behavior.  Testbenches, BFMs, models, etc. that run over time in a simulation usually should use `inject`.
